### ISSUE TITLE
`onbeforeremove` does not always trigger when navigating away

### ISSUE THREAD

**Issue by osban** [MithrilJS/mithril.js#2090]
In case of `Layout -> Parent -> Child` components, `onbeforeremove` in `Child` is triggered normally when clicking and navigating away through `Parent`, but not when clicking and navigating away through `Layout`.

## Expected Behavior
`onbeforeremove` should always trigger when navigating away from the component, `Child` in this case.

## Current Behavior
`onbeforeremove` in `Child` is triggered normally when clicking and navigating away through `Parent`, but not when clicking and navigating away through `Layout`.

## Steps to Reproduce (for bugs)
[Flems](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4hcYgAIAMhgCetAK5SAvJOAAdNJMkA3CDADuiPWloATGJJUA+Sci06dWABQBycxF1uKjp8-cMH3U-fx1CACcYMBM3AHofULD6aiiMRhMsfAilRnwoCDQAaySddkkNEAAJWhxKgEpfbTDXNyCKEOawyUjo2LjiDDgixK7ktFSYdJhM7NyYfMKSsbKOyoAVIZKQeqTGpNa4GGpiCHpg3TNLfGpCaHMotF3mgF0tVi0tcSkABQxH1SdHT0QoQYgmFzACTTVj1ax2KGDPKDYbWSRgDBQI7sPz6IwQxEwuG2ewHdyebxNbqtQgAJmClT+AIaVJa7gARspiOcOsAUgVqEUIcS7NDkVs0cQIooYOVKgBlQi0Qws0qSVqc4jctDBPkTAVCyQuEWSMULFFFNEYrGytbVCCWBp7Vk6M34C2SABknvVLgAwncoOZnjo3mgPmgvvQJJIA-c0ZpmvR2dFaFFFAAHczTYXw9QZlyVZOp9NZ6YNHFJtApsBpmBRLC0XQzI0m4AFovVkv1mCN5sV1l44ytvMOMatCmjanuOkMkBxoNOsYhyRhiNaLI5ZQwFzmWjURQ4Bj4dkWOQdeK6vyVOKVEzAR6WCK5kmuWQKZT1Ss6G8Wu-qR96xfOw33kXIOlcJkRHqL93nqNgOBATAcDwG44AEGh6EYZgeDYF4qAKYoEBQTgUJ4LAwUiaBKBARQInIHgSGIDM4EQOI4kUNAMyKABzG5ajiCjiCoqAAAEAEZ8EkgBWQTKIiaB8H4GjiDkDNuBAZT2FIjShJEgBaCQ0iwGi6IYkAmJYtiOK43j+KwOThIUqBDKlKYsAkqT8AABjiIz3KUjDVPUvAtJeVggA).

## Context
I want to remove temporary files in the `onbeforeremove`.

## Your Environment
* Version used: 1.1.6
* Browser Name and version: Firefox 58.0.2
* Operating System and version (desktop or mobile): desktop Windows 10

**Comment by isiahmeadows** [comment]
Note: [this is apparently documented as intended behavior](https://gitter.im/mithriljs/mithril.js?at=5a8ad86be4ff28713a9a4a2b), although I personally object to it as *seriously* odd behavior that violates user expectations.

@MithrilJS/collaborators WDYT about fixing this in v2? (It's a pretty straightforward fix for me to write.)

**Comment by barneycarroll** [comment]
> I personally object to it as _seriously_ odd behavior that violates user expectations.

@isiahmeadows the problem is that if this behaviour is implemented, all sorts of scenarios arise where it's impossible to reasonably infer straightforward expectations. I proposed this in #1182, where we discussed the implementation difficulties and potential for unforeseen complexity.

**Comment by isiahmeadows** [comment]
@barneycarroll 

> the problem is that if this behaviour is implemented, all sorts of scenarios arise where it's impossible to reasonably infer straightforward expectations. I proposed this in #1182, where we discussed the implementation difficulties and potential for unforeseen complexity.

I've had a couple years to study highly async stuff further (including async DOM manipulation, which I've studied for quite a while, and other batched async updates), and here's what I'd reasonably expect:

1. All hooks are invoked at once, starting from the child and up to the parent.
1. Nothing is removed from a parent until its `onbeforeremove` and all its children's all resolve.
1. Once everything has been awaited (or if nothing needed awaited), then remove the tree.
1. This works holistically. If a child resolves before its parent, the child waits until the parent resolves before it's properly detached.
1. `onremove` hooks are fired in a separate pass after all `onbeforeremove` hooks are passed.
1. After those are all fired, the parent and all its children are unlinked. (The [`TreeWalker` API](https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker) could simplify this for us *drastically*.
1. Any future `m.render` calls on a tree would be blocked on existing `onbeforeremove` hooks. (This would also resolve [this bug](https://github.com/MithrilJS/mithril.js/issues/1881), which we need to do regardless of the outcome of this one.)

We could also add an extra options argument to `m.redraw` ([related](https://github.com/MithrilJS/mithril.js/issues/2074)) to accept a `resolve` and `reject` argument, called on successful and failed application of a new tree. Although those wouldn't be applicable to `m.route`, it'd be applicable elsewhere (like in tests, where we could actually check for scheduled timings). We could also use those to provide hooks to make `m.redraw()` (async version) return a promise that resolves when finished and have `m.redraw.sync()` accept an extra options object to also expose this facility.

-----

Needless to say, async isn't exactly easy nor simple, but that's mostly at the theoretical level. The implementation is much harder to *plan* than it is to *write* - it's hard to get *right*, but it's not super *verbose* to get right, at least not significantly more than what exists in this instance.

If we agree to this, I'll try to commit to it personally what I can.

**Comment by isiahmeadows** [comment]
As an alternative to my steps, you could swap steps 1-6 out to just do this, which *could* align with some's expectations at the cost of a little extra performance:

1. Do these steps on each child and await them.
1. Call `onbeforeremove` and await the result.
1. Call `onremove`.
1. Remove the node.

Either route we take *needs* to be documented explicitly so we can tell users what is *supposed* to happen.

**Comment by barneycarroll** [comment]
@osban summarising what we discussed on Gitter, unless I'm mistaken, your underlying issue was adequately resolved by using `onremove` instead, which is fired for every node in a removed branch.

@isiahmeadows since your proposal is orthogonal to @osban's concerns but essentially deals with reviving the conversation in #1182, shall we close this issue and re-open that one?

**Comment by isiahmeadows** [comment]
@barneycarroll Probably not a bad idea, and done.

**Comment by isiahmeadows** [comment]
I'm torn on whether to leave this open, as it's a near-duplicate of #1182, just 1.5 years later.

**Comment by osban** [comment]
@barneycarroll Yes, that's correct.

**Comment by isiahmeadows** [comment]
Closing, since the original issue appears fixed.
